pkg/util/debouncer/debouncer.go

```diff
@@ -0,0 +1,311 @@
+package debouncer
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/instrument"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	ErrBufferFull = errors.New("debouncer buffer full")
+)
+
+type KeyFunc[T any] func(T) string
+type ProcessFunc[T any] func(context.Context, T) error
+type ErrorFunc[T any] func(T, error)
+
+type metrics struct {
+	itemsAddedCounter           prometheus.Counter
+	itemsDroppedCounter         prometheus.Counter
+	itemsProcessedCounter       prometheus.Counter
+	processingErrorsCounter     prometheus.Counter
+	processingDurationHistogram prometheus.Histogram
+}
+
+func newMetrics(reg prometheus.Registerer, name string) *metrics {
+	return &metrics{
+		itemsAddedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_items_added_total",
+			Help: "Total number of items added to the debouncer",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		itemsDroppedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_items_dropped_total",
+			Help: "Total number of items dropped due to a full buffer",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		itemsProcessedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_items_processed_total",
+			Help: "Total number of items processed by the debouncer",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		processingErrorsCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "debouncer_processing_errors_total",
+			Help: "Total number of errors during processing",
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+		processingDurationHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
+			Name:                            "debouncer_processing_duration_seconds",
+			Help:                            "Time taken to process items",
+			Buckets:                         instrument.DefBuckets,
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  160,
+			NativeHistogramMinResetDuration: time.Hour,
+			ConstLabels: prometheus.Labels{
+				"name": name,
+			},
+		}),
+	}
+}
+
+type DebouncerOpts[T any] struct {
+	BufferSize     int
+	KeyFunc        KeyFunc[T]
```

Comment:
From what I can see, we only need `KeyFunc` to get a key for debouncers map. If `T` was defined as `T comparable` instead of `T any`, we could actually use it as a key to the map directly, and we wouldn't need `KeyFunc` at all. WDYT?

Reply:
I think that's a good call, if you ever want to have something that's fancy as key, like containing a slice, you can just use the pointer as key. Will change it 👍

Reply:
Done 👍